walkeeper/src/callmemaybe.rs

```diff
@@ -0,0 +1,293 @@
+//!
+//!  Callmemaybe thread is responsible for periodically requesting
+//!  pageserver to initiate wal streaming.
+//!
+use crate::SafeKeeperConf;
+use anyhow::Result;
+use log::*;
+use std::collections::HashMap;
+use std::net::ToSocketAddrs;
+use std::sync::Mutex;
+use std::time::{Duration, Instant};
+use tokio::runtime;
+use tokio::sync::mpsc::Receiver;
+use tokio::task;
+use tokio_postgres::NoTls;
+use zenith_utils::zid::{ZTenantId, ZTimelineId};
+
+async fn request_callback(
+    ps_addr_str: String,
+    listen_pg_addr_str: String,
+    pageserver_auth_token: Option<String>,
+    timelineid: ZTimelineId,
+    tenantid: ZTenantId,
+) -> Result<()> {
+    // Connect to the pageserver.
+    let mut config = tokio_postgres::Config::new();
+    let ps_addr = ps_addr_str.to_socket_addrs().unwrap().next().unwrap();
+
+    config.host(&ps_addr.ip().to_string()).port(ps_addr.port());
+
+    if let Some(password) = pageserver_auth_token {
+        config.password(password);
+    }
+
+    debug!(
+        "callmemaybe request_callback Connecting to pageserver {}",
+        ps_addr
+    );
+    let (client, connection) = config.connect(NoTls).await?;
+
+    tokio::spawn(async move {
+        if let Err(e) = connection.await {
+            eprintln!("connection error: {}", e);
+        }
+    });
+
+    // Send callmemaybe request
+    let listen_pg_addr = listen_pg_addr_str
+        .to_socket_addrs()
+        .unwrap()
+        .next()
+        .unwrap();
+
+    let callme = format!(
+        "callmemaybe {} {} host={} port={} options='-c ztimelineid={} ztenantid={}'",
+        tenantid,
+        timelineid,
+        &listen_pg_addr.ip().to_string(),
+        listen_pg_addr.port(),
+        timelineid,
+        tenantid
+    );
+
+    let _ = client.simple_query(&callme).await?;
+
+    Ok(())
+}
+
+pub fn thread_main(conf: SafeKeeperConf, rx: Receiver<CallmeEvent>) -> Result<()> {
+    let runtime = runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()
+        .unwrap();
+
+    runtime.block_on(main_loop(conf, rx))
+}
+
+#[derive(Debug)]
+pub enum CallmeEvent {
+    // add new subscription to the list
+    Subscribe(ZTenantId, ZTimelineId),
+    // remove the subscription from the list
+    Unsubscribe(ZTenantId, ZTimelineId),
+    // don't serve this subscription, but keep it in the list
+    Pause(ZTenantId, ZTimelineId),
+    // resume this subscription, if it exists,
+    // but don't create a new one if it is gone
+    Resume(ZTenantId, ZTimelineId),
+}
+
+struct SubscriptionStateInner {
+    handle: Option<task::JoinHandle<()>>,
+    last_call_time: Instant,
+    paused: bool,
+}
+
+struct SubscriptionState {
+    tenantid: ZTenantId,
+    timelineid: ZTimelineId,
+    inner: Mutex<SubscriptionStateInner>,
```

Comment:
current_thread runtime is used here (and we don't need more than that), so there is nothing to block/being blocked by. It would be nice to drop the mutex altogether actually.

Reply:
If we hold the mutex across await points it might still make sense in a single threaded runtime. To actually spawn non send futures see [LocalSet](https://docs.rs/tokio/1.14.0/tokio/task/struct.LocalSet.html). Mutex might be replaced with Rc<RefCell>. But still beware of holding RefMut across await points, compiler should issue a warning in that case.